src/control/tools/SplineHandler.cpp

```diff
@@ -0,0 +1,275 @@
+        stroke = nullptr;
+        xournal->getCursor()->updateCursor();
+        return;
+    }
+
+    int dpiScaleFactor = xournal->getDpiScaleFactor();
+    cairo_scale(cr, zoom * dpiScaleFactor, zoom * dpiScaleFactor);
+
+    // draw circles around knot points
+    double lineWidth = 0.5 * zoom;
+    int pointCount = stroke->getPointCount();
+    cairo_set_line_width(cr, lineWidth);
+    Point firstPoint = stroke->getPoint(0);
+    Point currPoint = stroke->getPoint(pointCount - 1);
+    double dist = currPoint.lineLengthTo(firstPoint);
+    cairo_set_source_rgb(cr, 1, 0, 0);                               // use red color for first knot
+    cairo_move_to(cr, firstPoint.x + radius, firstPoint.y);          // move to start point of circle arc;
+    cairo_arc(cr, firstPoint.x, firstPoint.y, radius, 0, 2 * M_PI);  // draw circle
+    if (dist < radius && pointCount > 3) {  // current point lies within the circle around the first knot
+        cairo_fill(cr);
+    } else {
+        cairo_stroke(cr);
+    }
+
+    cairo_set_source_rgb(cr, 0.3, 0.3, 0.3);       // use gray color for all knots except first one
+    for (size_t i = 1; i < pointCount - 1; i++) {  // dynamically changing knot is not circled
+        Point p = stroke->getPoint(i);
+        cairo_move_to(cr, p.x + radius, p.y);          // move to start point of circle arc;
+        cairo_arc(cr, p.x, p.y, radius, 0, 2 * M_PI);  // draw circle
+    }
+    cairo_stroke(cr);
+
+    // draw spline
+    view.drawStroke(cr, stroke, 0);
+}
+
+auto SplineHandler::onKeyEvent(GdkEventKey* event) -> bool {
+    if (!stroke ||
+        event->type != GDK_KEY_PRESS && event->keyval != GDK_KEY_Escape) {  // except for escape key only act on key
+                                                                            // press event, not on key release event
+        return false;
+    }
+
+    int pointCount = stroke->getPointCount();
+    double zoom = xournal->getZoom();
+    double radius = radiusConst * zoom;
+    double x = stroke->getX() - radius;
+    double y = stroke->getY() - radius;
+    double w = stroke->getElementWidth() + 2 * radius;
+    double h = stroke->getElementHeight() + 2 * radius;
+
+    switch (event->keyval) {
+        case GDK_KEY_Escape: {
+            if (pointCount > 1) {
+                // remove dynamically changing point at cursor position
+                stroke->deletePoint(pointCount - 1);
+                this->redrawable->repaintRect(x, y, w, h);
+            }
+            finalizeSpline();
+            return true;
+        }
+        // Actions after backspace or key arrow press
+        case GDK_KEY_BackSpace: {
+            if (pointCount > 2) {
+                // delete last non dynammically changing point
+                stroke->deletePoint(pointCount - 2);
+                this->redrawable->repaintRect(x, y, w, h);
+            }
+            break;
+        }
+        case GDK_KEY_Up: {
+            if (pointCount > 2) {
```

Comment:
There's a lot of code duplication here. The common parts should be factored out into a helper function.

Reply:
I have now factored the common parts from the arrow key events into a helper function called movePoint